### PR TITLE
Update to using new project model introduced in 2021.1 EAP3

### PR DESCRIPTION
**Pull request by maartenba**
Updates to the project model. Tested everything on myend and seems OK, but feel free to do another run-through.

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/FunctionAppContextPublishProvider.kt [hunk 1]
```diff
@@ -50,10 +50,10 @@ class FunctionAppContextPublishProvider : RiderContextPublishProvider {
         get() = RUN_CONFIG_NAME
 
     override fun getConfigurationForNode(project: Project,
-                                         projectModelNode: ProjectModelNode): Pair<RunConfiguration, ConfigurationFactory> {
+                                         entity: ProjectModelEntity): Pair<RunConfiguration, ConfigurationFactory> {
 
-        val projectData = RiderContextPublishProvider.getProjectDataRecursive(project, projectModelNode)
-                ?: error("Unexpected project node type. Cannot get project data for node ${projectModelNode.location}")
+        val projectData = RiderContextPublishProvider.getProjectDataRecursive(project, entity)
+                ?: error("Unexpected project node type. Cannot get project data for entity ${entity.url}")
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/FunctionAppContextPublishProvider.kt:56]: New entity has `getFileLocation()` method (or `getVirtualFileLocation()`). Does `url` give the same result? Can it be null for local files?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt [hunk 2]
```diff
@@ -52,8 +53,8 @@ class AzureCoreToolsMissingNupkgInstaller : StartupActivity {
                 file.extension.equals("vb", true) ||
                 file.extension.equals("fs", true)
 
-        private fun existsInCurrentProject(projectFileIndex: ProjectFileIndex, file: VirtualFile): Boolean =
-                file.exists() && projectFileIndex.isInContent(file)
+        private fun existsInCurrentProject(project: Project, workspaceModel: WorkspaceModel, file: VirtualFile): Boolean =
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: You add `workspaceModel: WorkspaceModel` parameter here, but does not use it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt [hunk 3]
```diff
@@ -52,8 +53,8 @@ class AzureCoreToolsMissingNupkgInstaller : StartupActivity {
...
+                file.exists() && VfsUtil.isAncestor(VfsUtil.findFileByIoFile(File(project.basePath!!), false)!!, file, false)
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: Would not it be more clear to check with this API?:
```
workspaceModel.getProjectModelEntities(file, project).isNotEmpty()
```
I'm a bit concerned about NPEs with number of unwraps :)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt [hunk 4]
```diff
@@ -140,19 +141,21 @@ class AzureCoreToolsMissingNupkgInstaller : StartupActivity {
...
-        val projects: HashSet<ProjectModelNode> = hashSetOf()
+fun WorkspaceModel.findProjectsWithVirtualFile(virtualFile: VirtualFile): List<ProjectModelEntity> {
+    val visitor = object : ProjectModelEntityVisitor() {
+        val projects: HashSet<ProjectModelEntity> = hashSetOf()
 
-        override fun visitNode(node: ProjectModelNode): Result {
-            if (node.getVirtualFile() == virtualFile) {
-                node.containingProject()?.let { projects.add(it) }
+        override fun visitNode(entity: ProjectModelEntity): Result {
+            if (entity.url?.virtualFile == virtualFile) {
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: I did not test this, but I believe it is `getVirtualFileAsContentRoot()` API for this case.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/actions/RegisterApplicationInAzureAdAction.kt [hunk 5]
```diff
@@ -298,18 +299,18 @@ class RegisterApplicationInAzureAdAction
...
 
-    private fun tryGetAppSettingsJsonVirtualFile(item: ProjectModelNode): VirtualFile? {
-        val itemVirtualFile = item.getVirtualFile()
+    private fun tryGetAppSettingsJsonVirtualFile(entity: ProjectModelEntity): VirtualFile? {
+        val itemVirtualFile = entity.url?.virtualFile
 
         if (isAppSettingsJsonFileName(itemVirtualFile?.name)) return itemVirtualFile
 
-        return item.containingProject()?.getVirtualFile()?.parent?.findChild(appSettingsJsonFileName)
+        return entity.containingProjectEntity()?.url?.virtualFile?.parent?.findChild(appSettingsJsonFileName)
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/actions/RegisterApplicationInAzureAdAction.kt]: There is `getVirtualFileAsParent()` API (I did not check on my end).

**Comment by sdubov** [comment]
@xtmq, could you please make a quick check for new workspace model changes?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt (hunk 3 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: Unfortunately, `workspaceModel.getProjectModelEntities(file, project).isNotEmpty()` always returns `false` in these cases. The model only gets updated later.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/FunctionAppContextPublishProvider.kt (hunk 1 above)
Comment (xtmq) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/FunctionAppContextPublishProvider.kt:56]: Use url, it is OK. 
Have no idea about `getVirtualFileLocation` method although...

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt (hunk 3 above)
Comment (xtmq) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: This code would not work in general. 
The proper way: `WorkspaceProjectRootsTracker.getInstance(project).contains(file)`

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt [hunk 6]
```diff
@@ -107,19 +108,19 @@ class AzureCoreToolsMissingNupkgInstaller : StartupActivity {
                             // Check for known marker words
                             if (knownMarkerWords.any { text.contains(it, true) }) {
                                 // Determine project(s) to install into
-                                val installableProjects = project.getComponent<ProjectModelViewHost>()
-                                        .findProjectsWithVirtualFile(file)
+                                val installableProjects = workspaceModel.findProjectsWithVirtualFile(file)
```
Comment (xtmq) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt:111]: Where are `findProjects` and `findProjectsByPath` methods

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt (hunk 4 above)
Comment (xtmq) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: It seems we are looking for any project with specific files inside. I suggest write something like that: 
```
WorkspaceModel.getInstance().getProjectModelEntities(virtualFile, project).map { it.containingProjectEntity() }
```

Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: The file is not yet in the model at this point :(

Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: is there a way to "refresh" the model?

Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsMissingNupkgInstaller.kt]: Nevermind, found a way to make this all work. Updates incoming!

**Review by sdubov — APPROVED** [review]

**Pull request merged**